File: deploy/lite/android/sdk/src/main/java/com/baidu/paddlex/preprocess/Transforms.java
// Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.baidu.paddlex.preprocess;
import android.util.Log;
import org.opencv.android.OpenCVLoader;
import org.opencv.core.Core;
import org.opencv.core.CvType;
import org.opencv.core.Mat;
import org.opencv.core.Rect;
import org.opencv.core.Scalar;
import org.opencv.core.Size;
import org.opencv.imgproc.Imgproc;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;

public class Transforms {
    private static final String TAG = Transforms.class.getSimpleName();
    private List<transformOp> transformOps = new ArrayList<transformOp>();
    private String transformsMode = "RGB";
	private HashMap<String, Integer> interpMap = new HashMap<String, Integer>(){{
        put("LINEAR", Imgproc.INTER_LINEAR);
        put("NEAREST", Imgproc.INTER_NEAREST);
        put("AREA", Imgproc.INTER_AREA);
        put("CUBIC", Imgproc.INTER_CUBIC);
        put("LANCZOS4", Imgproc.INTER_LANCZOS4);
        }
    };

    public void loadConfig(List transforms_list, String transformsMode) {
        if (!OpenCVLoader.initDebug()) {
            Log.e(TAG,"OpenCV Loadding failed.");
        }
        this.transformsMode = transformsMode;
        for (int i = 0; i < transforms_list.size(); i++) {
            HashMap transform_op = (HashMap) (transforms_list.get(i));
            if (transform_op.containsKey("ResizeByShort")) {
                HashMap info = (HashMap) transform_op.get("ResizeByShort");
                ResizeByShort resizeByShort = new ResizeByShort();
                resizeByShort.max_size = (int)info.get("max_size");
                resizeByShort.short_size = (int)info.get("short_size");
                if (info.containsKey("interp")) {
                    resizeByShort.interp = (String) info.get("interp");
                }
                transformOps.add(resizeByShort);
            } else if (transform_op.containsKey("ResizeByLong")) {
                HashMap info = (HashMap) transform_op.get("ResizeByLong");
                ResizeByLong resizeByLong = new ResizeByLong();
                resizeByLong.long_size = (int)info.get("long_size");
                if (info.containsKey("interp")) {
                    resizeByLong.interp = (String) info.get("interp");
                }
                transformOps.add(resizeByLong);

            } else if (transform_op.containsKey("CenterCrop")) {
                HashMap info = (HashMap) transform_op.get("CenterCrop");
                CenterCrop centerCrop = new CenterCrop();
                if (info.get("crop_size") instanceof Integer) {
                    centerCrop.cropHeight = (int) info.get("crop_size");
                    centerCrop.cropWidth = (int) info.get("crop_size");
                } else {
                    centerCrop.cropWidth = ((List<Integer>) info.get("crop_size")).get(0);
                    centerCrop.cropHeight = ((List<Integer>) info.get("crop_size")).get(1);
                }
                transformOps.add(centerCrop);
            } else if (transform_op.containsKey("Normalize")) {
                HashMap<String, List<Float>> info = (HashMap<String, List<Float>>) transform_op.get("Normalize");
                Normalize normalize = new Normalize();
                normalize.mean = info.get("mean").toArray(new Double[info.get("mean").size()]);
                normalize.std = info.get("std").toArray(new Double[info.get("std").size()]);
                transformOps.add(normalize);
            } else if (transform_op.containsKey("Resize")) {
                HashMap info = (HashMap) transform_op.get("Resize");
                Resize resize = new Resize();
                if (info.get("target_size") instanceof Integer) {
                    resize.width = (int) info.get("target_size");
                    resize.height = (int) info.get("target_size");
                } else {
                    resize.width = ((List<Integer>) info.get("target_size")).get(0);
                    resize.height = ((List<Integer>) info.get("target_size")).get(1);
                }
                if (info.containsKey("interp")) {
                    resize.interp = (String) info.get("interp");
                }
                transformOps.add(resize);
            } else if (transform_op.containsKey("Padding")) {
                HashMap info = (HashMap) transform_op.get("Padding");
                Padding padding = new Padding();
                if (info.containsKey("coarsest_stride")) {
                    padding.coarsest_stride = (int) info.get("coarsest_stride");
                }
                if (info.containsKey("im_padding_value")) {
                    List<Double> im_padding_value = (List<Double>) info.get("im_padding_value");
                    if (im_padding_value.size()!=3){
                        Log.e(TAG, "len of im_padding_value in padding must == 3.");
                    }
                    for (int k =0; i<im_padding_value.size(); i++){
                        padding.paddding_value[k] = im_padding_value.get(k);
                    }
                }
                if (info.containsKey("target_size")) {
                    if (info.get("target_size") instanceof Integer) {
                        padding.width = (int) info.get("target_size");
                        padding.height = (int) info.get("target_size");
                    } else {
                        padding.width = ((List<Integer>) info.get("target_size")).get(0);
                        padding.height = ((List<Integer>) info.get("target_size")).get(1);
                    }
                }
                transformOps.add(padding);
            }
        }
    }

    public ImageBlob run(Mat inputMat, ImageBlob imageBlob) {
        imageBlob.setOriImageSize(inputMat.height(),2);
        imageBlob.setOriImageSize(inputMat.width(),3);
        imageBlob.setNewImageSize(inputMat.height(),2);
        imageBlob.setNewImageSize(inputMat.width(),3);

        if(transformsMode.equalsIgnoreCase("RGB")){
            Imgproc.cvtColor(inputMat, inputMat, Imgproc.COLOR_BGR2RGB);
        }else if(!transformsMode.equalsIgnoreCase("BGR")){
            Log.e(TAG, "transformsMode only support RGB or BGR.");
        }
        inputMat.convertTo(inputMat, CvType.CV_32FC(3));

        for (transformOp op : transformOps) {
            inputMat = op.run(inputMat, imageBlob);
        }

        int w = inputMat.width();
        int h = inputMat.height();
        int c = inputMat.channels();
        imageBlob.setImageData(new float[w * h * c]);

        Mat singleChannelMat = new Mat(h, w, CvType.CV_32FC(1));
        float[] singleChannelImageData = new float[w * h];
        for (int i = 0; i < c; i++) {
            Core.extractChannel(inputMat, singleChannelMat, i);
            singleChannelMat.get(0, 0, singleChannelImageData);
            System.arraycopy(singleChannelImageData ,0, imageBlob.getImageData(),i*w*h, w*h);
        }

        return imageBlob;
    }

    private class transformOp {
        public Mat run(Mat inputMat, ImageBlob data) {
            return inputMat;
        }
    }

    private class ResizeByShort extends transformOp {
        private int max_size;
        private int short_size;
        private String interp = "LINEAR";

        public Mat run(Mat inputMat, ImageBlob imageBlob) {
            int origin_w = inputMat.width();
            int origin_h = inputMat.height();
            imageBlob.getReshapeInfo().put("resize", new int[]{origin_w, origin_h});
            int im_size_max = Math.max(origin_w, origin_h);
            int im_size_min = Math.min(origin_w, origin_h);
            float scale = (float) (short_size) / (float) (im_size_min);
            if (max_size > 0) {
                if (Math.round(scale * im_size_max) > max_size) {
                    scale = (float) (max_size) / (float) (im_size_max);
                }
            }
            int width = Math.round(scale * origin_w);
            int height = Math.round(scale * origin_h);
            Size sz = new Size(width, height);
            Imgproc.resize(inputMat, inputMat, sz,0,0, interpMap.get(interp));
            imageBlob.setNewImageSize(inputMat.height(),2);
            imageBlob.setNewImageSize(inputMat.width(),3);
            imageBlob.setScale(scale);
            return inputMat;
        }
    }

    private class ResizeByLong extends transformOp {
        private int long_size;
        private String interp = "LINEAR";

        public Mat run(Mat inputMat, ImageBlob imageBlob) {
            int origin_w = inputMat.width();
            int origin_h = inputMat.height();
            imageBlob.getReshapeInfo().put("resize", new int[]{origin_w, origin_h});
            int im_size_max = Math.max(origin_w, origin_h);
            float scale = (float) (long_size) / (float) (im_size_max);
            int width = Math.round(scale * origin_w);
            int height = Math.round(scale * origin_h);
            Size sz = new Size(width, height);
            Imgproc.resize(inputMat, inputMat, sz,0,0, interpMap.get(interp));
            imageBlob.setNewImageSize(inputMat.height(),2);
            imageBlob.setNewImageSize(inputMat.width(),3);
            imageBlob.setScale(scale);
            return inputMat;
        }
    }

    private class CenterCrop extends transformOp {
        private int cropHeight;
        private int cropWidth;

        public Mat run(Mat inputMat, ImageBlob imageBlob) {
            int origin_w = inputMat.width();
            int origin_h = inputMat.height();
            if (origin_h < cropHeight || origin_w < cropWidth) {
                Log.e(TAG, "[CenterCrop] Image size less than crop size");
            }
            int offset_x, offset_y;
            offset_x = (origin_w - cropWidth) / 2;
            offset_y = (origin_h - cropHeight) / 2;
            offset_x = Math.max(Math.min(offset_x, origin_w - cropWidth), 0);
            offset_y = Math.max(Math.min(offset_y, origin_h - cropHeight), 0);
            Rect crop_roi = new Rect(offset_x, offset_y, cropHeight, cropWidth);
            inputMat = inputMat.submat(crop_roi);
            imageBlob.setNewImageSize(inputMat.height(),2);
            imageBlob.setNewImageSize(inputMat.width(),3);
            return inputMat;
        }
    }

    private class Resize extends transformOp {
        private int height;
        private int width;
        private String interp = "LINEAR";

        public Mat run(Mat inputMat, ImageBlob imageBlob) {
            int origin_w = inputMat.width();
            int origin_h = inputMat.height();
            imageBlob.getReshapeInfo().put("resize", new int[]{origin_w, origin_h});
            Size sz = new Size(width, height);
            Imgproc.resize(inputMat, inputMat, sz,0,0,  interpMap.get(interp));
            imageBlob.setNewImageSize(inputMat.height(),2);
            imageBlob.setNewImageSize(inputMat.width(),3);
            return inputMat;
        }
    }

    private class Padding extends transformOp {
        private double width;
        private double height;
        private double coarsest_stride;
        private double[] paddding_value = {0.0, 0.0, 0.0};

        public Mat run(Mat inputMat, ImageBlob imageBlob) {
            int origin_w = inputMat.width();
            int origin_h = inputMat.height();
            imageBlob.getReshapeInfo().put("padding", new int[]{origin_w, origin_h});
            double padding_w = 0;
            double padding_h = 0;
            if (width > 1 & height > 1) {
                padding_w = width;
                padding_h = height;
            } else if (coarsest_stride > 1) {
                padding_h = Math.ceil(origin_h / coarsest_stride) * coarsest_stride;
                padding_w = Math.ceil(origin_w / coarsest_stride) * coarsest_stride;
            }
            imageBlob.setNewImageSize(inputMat.height(),2);
            imageBlob.setNewImageSize(inputMat.width(),3);
            Core.copyMakeBorder(inputMat, inputMat, 0, (int)padding_h, 0, (int)padding_w, Core.BORDER_CONSTANT, new Scalar(paddding_value));
            return inputMat;
        }
    }

    private class Normalize extends transformOp {
        private Double[] mean = new Double[3];
        private Double[] std = new Double[3];

        public Mat run(Mat inputMat, ImageBlob imageBlob) {
            inputMat.convertTo(inputMat, CvType.CV_32FC(3), 1/255.0);
            Scalar meanScalar = new Scalar(mean[0], mean[1], mean[2]);
            Scalar stdScalar = new Scalar(std[0], std[1], std[2]);
            Core.subtract(inputMat, meanScalar, inputMat);
            Core.divide(inputMat, stdScalar, inputMat);
            return inputMat;
        }
    }
}

File: deploy/cpp/src/transforms.cpp
//   Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <iostream>
#include <string>
#include <vector>

#include "include/paddlex/transforms.h"

namespace PaddleX {

std::map<std::string, int> interpolations = {{"LINEAR", cv::INTER_LINEAR},
                                             {"NEAREST", cv::INTER_NEAREST},
                                             {"AREA", cv::INTER_AREA},
                                             {"CUBIC", cv::INTER_CUBIC},
                                             {"LANCZOS4", cv::INTER_LANCZOS4}};

bool Normalize::Run(cv::Mat* im, ImageBlob* data) {
  for (int h = 0; h < im->rows; h++) {
    for (int w = 0; w < im->cols; w++) {
      im->at<cv::Vec3f>(h, w)[0] =
          (im->at<cv::Vec3f>(h, w)[0] / 255.0 - mean_[0]) / std_[0];
      im->at<cv::Vec3f>(h, w)[1] =
          (im->at<cv::Vec3f>(h, w)[1] / 255.0 - mean_[1]) / std_[1];
      im->at<cv::Vec3f>(h, w)[2] =
          (im->at<cv::Vec3f>(h, w)[2] / 255.0 - mean_[2]) / std_[2];
    }
  }
  return true;
}

float ResizeByShort::GenerateScale(const cv::Mat& im) {
  int origin_w = im.cols;
  int origin_h = im.rows;
  int im_size_max = std::max(origin_w, origin_h);
  int im_size_min = std::min(origin_w, origin_h);
  float scale =
      static_cast<float>(short_size_) / static_cast<float>(im_size_min);
  if (max_size_ > 0) {
    if (round(scale * im_size_max) > max_size_) {
      scale = static_cast<float>(max_size_) / static_cast<float>(im_size_max);
    }
  }
  return scale;
}

bool ResizeByShort::Run(cv::Mat* im, ImageBlob* data) {
  data->im_size_before_resize_.push_back({im->rows, im->cols});
  data->reshape_order_.push_back("resize");

  float scale = GenerateScale(*im);
  int width = static_cast<int>(scale * im->cols);
  int height = static_cast<int>(scale * im->rows);
  cv::resize(*im, *im, cv::Size(width, height), 0, 0, cv::INTER_LINEAR);

  data->new_im_size_[0] = im->rows;
  data->new_im_size_[1] = im->cols;
  data->scale = scale;
  return true;
}

bool CenterCrop::Run(cv::Mat* im, ImageBlob* data) {
  int height = static_cast<int>(im->rows);
  int width = static_cast<int>(im->cols);
  if (height < height_ || width < width_) {
    std::cerr << "[CenterCrop] Image size less than crop size" << std::endl;
    return false;
  }
  int offset_x = static_cast<int>((width - width_) / 2);
  int offset_y = static_cast<int>((height - height_) / 2);
  cv::Rect crop_roi(offset_x, offset_y, width_, height_);
  *im = (*im)(crop_roi);
  data->new_im_size_[0] = im->rows;
  data->new_im_size_[1] = im->cols;
  return true;
}

bool Padding::Run(cv::Mat* im, ImageBlob* data) {
  data->im_size_before_resize_.push_back({im->rows, im->cols});
  data->reshape_order_.push_back("padding");

  int padding_w = 0;
  int padding_h = 0;
  if (width_ > 1 & height_ > 1) {
    padding_w = width_ - im->cols;
    padding_h = height_ - im->rows;
  } else if (coarsest_stride_ >= 1) {
    int h = im->rows;
    int w = im->cols;
    padding_h =
        ceil(h * 1.0 / coarsest_stride_) * coarsest_stride_ - im->rows;
    padding_w =
        ceil(w * 1.0 / coarsest_stride_) * coarsest_stride_ - im->cols;
  }

  if (padding_h < 0 || padding_w < 0) {
    std::cerr << "[Padding] Computed padding_h=" << padding_h
              << ", padding_w=" << padding_w
              << ", but they should be greater than 0." << std::endl;
    return false;
  }
  cv::Scalar value = cv::Scalar(im_value_[0], im_value_[1], im_value_[2]);
  cv::copyMakeBorder(
      *im, *im, 0, padding_h, 0, padding_w, cv::BORDER_CONSTANT, value);
  data->new_im_size_[0] = im->rows;
  data->new_im_size_[1] = im->cols;
  return true;
}

bool ResizeByLong::Run(cv::Mat* im, ImageBlob* data) {
  if (long_size_ <= 0) {
    std::cerr << "[ResizeByLong] long_size should be greater than 0"
              << std::endl;
    return false;
  }
  data->im_size_before_resize_.push_back({im->rows, im->cols});
  data->reshape_order_.push_back("resize");
  int origin_w = im->cols;
  int origin_h = im->rows;

  int im_size_max = std::max(origin_w, origin_h);
  float scale =
      static_cast<float>(long_size_) / static_cast<float>(im_size_max);
  cv::resize(*im, *im, cv::Size(), scale, scale, cv::INTER_NEAREST);
  data->new_im_size_[0] = im->rows;
  data->new_im_size_[1] = im->cols;
  data->scale = scale;
  return true;
}

bool Resize::Run(cv::Mat* im, ImageBlob* data) {
  if (width_ <= 0 || height_ <= 0) {
    std::cerr << "[Resize] width and height should be greater than 0"
              << std::endl;
    return false;
  }
  if (interpolations.count(interp_) <= 0) {
    std::cerr << "[Resize] Invalid interpolation method: '" << interp_ << "'"
              << std::endl;
    return false;
  }
  data->im_size_before_resize_.push_back({im->rows, im->cols});
  data->reshape_order_.push_back("resize");

  cv::resize(
      *im, *im, cv::Size(width_, height_), 0, 0, interpolations[interp_]);
  data->new_im_size_[0] = im->rows;
  data->new_im_size_[1] = im->cols;
  return true;
}

void Transforms::Init(const YAML::Node& transforms_node, bool to_rgb) {
  transforms_.clear();
  to_rgb_ = to_rgb;
  for (const auto& item : transforms_node) {
    std::string name = item.begin()->first.as<std::string>();
    std::cout << "trans name: " << name << std::endl;
    std::shared_ptr<Transform> transform = CreateTransform(name);
    transform->Init(item.begin()->second);
    transforms_.push_back(transform);
  }
}

std::shared_ptr<Transform> Transforms::CreateTransform(
    const std::string& transform_name) {
  if (transform_name == "Normalize") {
    return std::make_shared<Normalize>();
  } else if (transform_name == "ResizeByShort") {
    return std::make_shared<ResizeByShort>();
  } else if (transform_name == "CenterCrop") {
    return std::make_shared<CenterCrop>();
  } else if (transform_name == "Resize") {
    return std::make_shared<Resize>();
  } else if (transform_name == "Padding") {
    return std::make_shared<Padding>();
  } else if (transform_name == "ResizeByLong") {
    return std::make_shared<ResizeByLong>();
  } else {
    std::cerr << "There's unexpected transform(name='" << transform_name
              << "')." << std::endl;
    exit(-1);
  }
}

bool Transforms::Run(cv::Mat* im, ImageBlob* data) {
  // 按照transforms中预处理算子顺序处理图像
  if (to_rgb_) {
    cv::cvtColor(*im, *im, cv::COLOR_BGR2RGB);
  }
  (*im).convertTo(*im, CV_32FC3);
  data->ori_im_size_[0] = im->rows;
  data->ori_im_size_[1] = im->cols;
  data->new_im_size_[0] = im->rows;
  data->new_im_size_[1] = im->cols;
  for (int i = 0; i < transforms_.size(); ++i) {
    if (!transforms_[i]->Run(im, data)) {
      std::cerr << "Apply transforms to image failed!" << std::endl;
      return false;
    }
  }

  // 将图像由NHWC转为NCHW格式
  // 同时转为连续的内存块存储到ImageBlob
  int h = im->rows;
  int w = im->cols;
  int c = im->channels();
  (data->im_data_).resize(c * h * w);
  float* ptr = (data->im_data_).data();
  for (int i = 0; i < c; ++i) {
    cv::extractChannel(*im, cv::Mat(h, w, CV_32FC1, ptr + i * h * w), i);
  }
  return true;
}

}  // namespace PaddleX
